### ISSUE TITLE
Recommend fix for Header Injection, Correct Error with Public Website Team

### ISSUE THREAD

**Issue by KevinMHoffmanUSDS** [department-of-veterans-affairs/va.gov-team#6155]
## What happened?
Header injection script is rendering improperly on IE 11 on GFE (government furnished equipment)

Background: https://github.com/department-of-veterans-affairs/va.gov-team/search?q=header+injection&type=Issues

## Specs:
Internet Explorer 11
Operating system Windows NT 10.0
IP address—
help_outline
Javascript enabled Yes
Cookies enabled Yes
Flash version 32.0.0
Java version 1.8.0_231
Websockets supported Yes
WebGL supported Yes
Language en-US
Browser size 1452 x 454
Screen size 1536 x 864 (Retina)
Color depth 24 bit
AdBlock enabled No

## Steps to Reproduce
- URL: https://www.oprm.va.gov/

## Desired behavior
Header and footer should render properly. 

## Acceptance Criteria
- [ ] Clearly define the cause of the bug and recommend a fix to the public website team. 
- [ ] Partner with a developer on the public website to fix the bug. 
- [ ] Implement the fix. 

## How to configure this issue
- [ ] **Labeled with Team** (`tools-improvements`)
- [ ] **Labeled with Practice Area** (`backend`, `frontend`)
- [ ] **Labeled with `Bug`**

**Comment by KevinMHoffmanUSDS** [comment]
@brandonrapp Hey Brandon! You haven't had a chance to meet with @drorva yet, but he may already have a fix for this. Could you coordinate with an FE on your team to see if that fix is indeed the right one, and get this to the public websites team this week? Let me know if you have any questions at all.

**Comment by ncksllvn** [comment]
Screenshot of https://www.oprm.va.gov/ using IE11 -

![image](https://user-images.githubusercontent.com/1915775/75264693-c85a0680-57bd-11ea-8409-0fef97ed66a9.png)

**Comment by ncksllvn** [comment]
I suspect that a runtime error is occurring mid-way through the injection process.

**Comment by ncksllvn** [comment]
Footer -

![image](https://user-images.githubusercontent.com/1915775/75264764-e293e480-57bd-11ea-95c7-1b04d2b19f3d.png)

**Comment by cvalarida** [comment]
I don't have a GFE to test on myself, but using IE11 in BrowserStack, it looks fine. :confused: 

## oprm.va.gov
### Header
![image](https://user-images.githubusercontent.com/12970166/75278598-099ee600-57bf-11ea-83d1-2213763a72ab.png)

### Header expanded
![image](https://user-images.githubusercontent.com/12970166/75278578-fe4bba80-57be-11ea-87f0-84cb43ea6c90.png)

### Footer
![image](https://user-images.githubusercontent.com/12970166/75278625-16bbd500-57bf-11ea-95d5-3e95a69de730.png)

## hiv.va.gov
### Header
![image](https://user-images.githubusercontent.com/12970166/75278708-38b55780-57bf-11ea-8ccc-558453e01212.png)

### Footer
![image](https://user-images.githubusercontent.com/12970166/75278733-41a62900-57bf-11ea-8337-64d1e5f59b50.png)

---

@ncksllvn Any idea what the difference might be between BrowserStack and the GFE reported in this issue?

**Comment by drorva** [comment]
Are there any errors on the console in browserstack?

**Comment by ncksllvn** [comment]
Hmmm. So Browser Stack seems okay then....so it has to be something unique about government-issued laptops on IE11. We may have to do a screenshare session with someone on a VA laptop in order to debug this. There may be a console error.

I know that VA-issued laptops require the `www` when navigating to `www.va.gov`, but I don't see how that would have any impact here.

**Comment by cvalarida** [comment]
@drorva Huh. Only when I open the dev tools and refresh the page, but turns out there _is_ an error. I just don't see the effect of it when I load up the page in IE 11. :confused: 

![image](https://user-images.githubusercontent.com/12970166/75365516-26482600-5872-11ea-8029-7abbbd7823d2.png)

![image](https://user-images.githubusercontent.com/12970166/75365393-fdc02c00-5871-11ea-8c73-78c027f86367.png)

That's...mildly confusing, though, since I'm not seeing any `.matches` call, but...hey, it's something...I guess. :sweat_smile:

**Comment by ncksllvn** [comment]
There is a call to `node.matches` on line 53 of `proxy-rewrite-entry`. It looks like IE uses a non-standard name - https://developer.mozilla.org/en-US/docs/Web/API/Element/matches - `msMatchesSelector`.

**Comment by ncksllvn** [comment]
Still a little confusing, because if that is the issue, it seems like this issue would be evident on all inject headers using Internet Explorer, not just this domain or on government-issued laptops.

**Comment by ncksllvn** [comment]
But an easy fix for this seems to be just checking for whether we should use `node.matches` or `node. msMatchesSelector`

**Comment by cvalarida** [comment]
Or polyfill `.matches` in IE11? Not sure what the level of effort on that would be like offhand.

Do you have a way to recreate the bug, @ncksllvn? Was that screenshot you posted above one you took?

**Comment by ncksllvn** [comment]
Someone on the content team provided me with the screenshot. I would have to use Browser Stack or something similar.

**Comment by AkucherGCIO** [comment]
2020-26-02 PW Planning 
@ncksllvn  to pair with on Engineer on Platform Team to transform knowledge on header injection fix.  

 @KevinMHoffmanUSDS @drorva  - Should Platform Team handle this type of work going forward ?

**Comment by cvalarida** [comment]
Just verified on my personal Windows 10 machine that the result was the same as I found in BrowserStack. Ran into that `.matches` error, but it didn't seem to affect anything.

Also discovered that IE11 gets a little annoyed at `<!-- HTML comments -->` too. :roll_eyes:

I'll see if I can get an engineer with a GFE to reproduce the bug.

**Comment by drorva** [comment]
![capture](https://user-images.githubusercontent.com/61249448/75473609-7a7f0300-5963-11ea-963d-73ea7f65c0b6.png)

This is from my GFE. As you can see the error is 
```
polyfills.entry.js (1,1183)
```

And a google search leads to https://github.com/babel/babel-loader/issues/401/ which makes me think that the issue might be that we load the polyfill library twice.

More importantly, if this is a GFE only issue, and it works fine in other WIN10 computers, my inclination is to mark the priority on this as low, and move it to the backlog. But this is a PM decision based on the size of the affected devices.

**Comment by cvalarida** [comment]
Ahhh, compatibility view! :male_detective: That's a different story now. :smile: 

When I add `va.gov` to the list of websites to use "compatibility view" on, I can reproduce the bug.
![image](https://user-images.githubusercontent.com/12970166/75476171-fcb0fc80-594e-11ea-9d70-41f853561bd8.png)

I don't _think_ the issue is that we're loading the polyfill library twice. We're seeing different errors than in that babel-loader issue. It'll take more digging into, but now that we can consistently reproduce the error, we have a path forward. :tada:

**Comment by KevinMHoffmanUSDS** [comment]
@AkucherGCIO Good question - I'd say there's a bit of history here, but going forward, if it's an issue with the header or footer code ITSELF, that would be the public website team, because that team owns those "products." If it's an issue with the way that code is being served to other sites, or what sites to serve it to, that would be the platform team. @jenniferlee-dsva discussed this this morning, and because we're unsure which of these is actually happening, I proposed that we touch base.

**Comment by cvalarida** [comment]
Further observations about the bug:

1. It's not present on www.va.gov using IE11 in compatibility mode
    - It's only the teamsite header / footer injection, it would seem
2. The `polyfills.entry.js` is loaded only once
    - On www.va.gov, we're loading `polyfills.entry.<hash>.js`
    - The `polyfills.entry.js` has a sourcemaps reference at the bottom of the file that points to the same hash as the file loaded on www.va.gov
3. www.oprm.va.gov is loading up a lot more JS files than www.va.gov
    - @drorva's hunch about the polyfills might be right if it's being included in one of those
    - None of them _look_ like they would conflict, but it's hard to say
    - The JS files from our S3 bucket are requested (and loaded) first
      - Makes sense since we're injecting them into the `<head>`
4. The errors I'm getting don't seem to match up with the source mapped version of the file
    - Getting a lot of `Expected identifier` errors, but the source maps are taking me to the wrong file altogether :shakes-fist:

So the good news is our scope is pretty narrow. I still don't know the _cause_ since I'm not familiar with what compatibility view even does, which means I still don't know who should tackle the fix. :sweat_smile:

**Comment by drorva** [comment]
At this point, my inclination is to either move it to the back log or even mark it as wontfix since it impacts a small segment of users. @KevinMHoffmanUSDS , you talked to the reporter, agreed?

**Comment by ncksllvn** [comment]
Compatibility view in IE is supposed to be a dev tool so that devs can use IE 11 to get an idea of what the page looks like on older versions of IE. Something like that. Of course, the compatibility view itself isn't perfect, so can't really be considered 100% reliable.

**Comment by ncksllvn** [comment]
I opened a PR that polyfills the `element.matches` method - https://github.com/department-of-veterans-affairs/vets-website/pull/11796 but my confidence that it actually fixes this issue is at about 10%.

Can someone take a look at that PR and try testing it out locally?

**Comment by KevinMHoffmanUSDS** [comment]
@drorva I wouldn't move this until the backlog until we are able to tell the team what is causing the issue. Do you have insight into that? My understanding from the above is that this is a javascript conflict caused by additional javascript files that they are loading. If someone could write this up, giving them a sense of the problem. This seems like the write up to me, is this correct? 

www.oprm.va.gov is loading up a lot more JS files than www.va.gov
@drorva's hunch about the polyfills might be right if it's being included in one of those
None of them look like they would conflict, but it's hard to say
The JS files from our S3 bucket are requested (and loaded) first
Makes sense since we're injecting them into the <head>

**Comment by AkucherGCIO** [comment]
03/04/20 PW Standup
Nick does not have a government issued laptop or windows machine, therefore is unable to test this fix.  A PR is open for this that Nick thinks will fix the issue.
@drorva , @jenniferlee-dsva - FYSA

**Comment by Cooperjstn** [comment]
Hello Kevin, this is Justin Cooper, (Contractor) Web Application Developer for OPRM, the site you are currently working on in regards to the injection issue/header/footer. VA Stakeholder would like to know if in the meantime we should create banner to notify visitors of site if they should view OPRM website using Google Chrome? Please advise.

**Comment by KevinMHoffmanUSDS** [comment]
Hello, @Cooperjstn - based on the above, it looks like we have a fix underway, but in the meantime, that seems like a reasonable suggestion.

**Comment by Cooperjstn** [comment]
Thank you @KevinMHoffmanUSDS for your help in regards to the issue. We appreciate your hard work in regards to this issue.

**Comment by jenniferlee-dsva** [comment]
Thank you @KevinMHoffmanUSDS @drorva @ncksllvn, and @AkucherGCIO for following.

**Comment by AkucherGCIO** [comment]
@ncksllvn - Has the PR been merged and have we been able to have anyone test this issue ?  If so, can we close this as part of Sprint 16 ?

**Comment by ncksllvn** [comment]
I opened that PR but have low confidence that it'll actually fix the issue and don't have the equipment to test the solution or replicate the error. I'll reach out to the other FE team members to see if someone will help me proceed.

**Comment by AkucherGCIO** [comment]
03/11/20 Sprint 17 Planning
@drorva @KevinMHoffmanUSDS - Isthis something that needs to be fixed on the VA side or is this FE Engineering on the Public Websites side ?  Last conversation, it sounded like it was something that had to happen on the VA Network side but we would like to confirm before we close this ticket.

**Comment by ncksllvn** [comment]
I'm able to produce a runtime error using Internet Explorer 10 via a cloud tool to run a Windows box. But we don't officially support IE 10, so could be a hard to do, and the issue clearly says IE 11. I wonder if there is a Compatibility View enabled in IE on GTE?

![image.png](https://images.zenhubusercontent.com/59cb9de1b0222d5de47953d8/b41099ed-e8cb-48f8-ba34-a27f6deb72d2)

**Comment by drorva** [comment]
Can we easily look up the % of users/month that use this kind of configuration. Let's not do that if it's complicated (but I'd like to know). 
Do we have policies about how small a percentage and support requirement?

**Comment by cvalarida** [comment]
Looking in Google Analytics, I'm seeing the following:

**Top 6 Browsers**
![image](https://user-images.githubusercontent.com/12970166/76462741-68f41d00-639f-11ea-9ad4-e0371fb89cfc.png)

**Internet Explorer Versions**
![image](https://user-images.githubusercontent.com/12970166/76462623-38ac7e80-639f-11ea-9170-5c701aa51829.png)

Of the 851k total users, IE @ < v11 represents ~3.3% of our user base. :grimacing: 
`(28186+228+67+9+1)÷851199 = 0.033471609`

I don't know if this number reflects IE 11 in compatibility view, but if not, that percentage only gets bigger.

To be clear, I expect this is a problem with IE 11 in compatibility view _and_ IE < 11. Not sure if I can verify that, though. (Can I run IE < 11 in Windows 10? :thinking: )

All that said, I'm _suuuuper_ interested in where all those IE7s are coming from! :male_detective: More people are on VA.gov using IE7 than all versions of Firefox! :dizzy_face:

**Comment by KevinMHoffmanUSDS** [comment]
@drorva Just FYI, this issue was revealed by a VA employee viewing on VA software on VA machine through the VA network.

**Comment by KevinMHoffmanUSDS** [comment]
@AkucherGCIO The goal of this ticket is to recommend a solution. Do we have a clear solution? Based on the solution recommended, we should probably know enough to know which team should solve the problem (e.g. if this is a front end code issue or a network issue.)

**Comment by AkucherGCIO** [comment]
@KevinMHoffmanUSDS - @ncksllvn has opened a PR with the fix, but was not able to test it and therefore can't say if that is a fix or not.  Maybe we have a quick sync up on how we can get someone to help test the fix ?

**Comment by cvalarida** [comment]
I tested the fix and commented in the PR. It didn't work, unfortunately.

To be clear, this a front end code issue, not a VA network. The trouble is on versions of IE that we don't support (and IE 11 simulating versions we don't support).

Where the VA machines might play tricks on us, however, is if that `Display intranet sites in Compatibility View` checkbox is checked _and_ machines on the VA network treat `va.gov` sites as intranet sites.

Something about the way our code interacts with the other JS on Teamsite sites is causing some weird things to happen, as @KevinMHoffmanUSDS mentioned. @drorva suggests it might be the polyfills being included twice (we include it once, the Teamsite site includes it once), which is a possibility, but I suspect that stacktrace is somewhat misleading. Just a hunch, though.

**Comment by KevinMHoffmanUSDS** [comment]
@cvalarida @ncksllvn So I want to make sure I reflect this correctly; the issue is with javascript that exists in the Teamsites, correct? Is there a way that @drorva (who has GFE with this browser) can partner with you to confirm WHICH javascript is causing the problem, so that we can advise them on how to fix?

**Comment by drorva** [comment]
/1. We need to have a clear policy about which browsers and version we support. I saw last night a discussion of this https://app.slack.com/client/T03FECE8V/CJRQ85PQB/thread/CJRQ85PQB-1583937995.197100. @cvalarida the stats from analytics are very helpful. I'm pretty clear that if a bug affects 3% of our visitors, we need to fix it, but it also depends on severity. If this is just cosmetic, I'm willing to let it go on old browsers. 

2. We need to be able to debug/replicate this. Browserstack doesn't have versions prior to IE11? I On one hand I'm totally happy to help with this, but on the other hand, this is not a long term solution.

**Comment by ncksllvn** [comment]
Sorry, this thread is hard to follow and define next steps, but like @cvalarida said -

> Where the VA machines might play tricks on us, however, is if that Display intranet sites in Compatibility View checkbox is checked and machines on the VA network treat va.gov sites as intranet sites.

This is what my gut is telling me. This issue is _not_ tied only to GFE laptops. It can be reproduced by anyone running IE11 in [compatibility view](https://winaero.com/blog/how-to-enable-compatibility-view-in-internet-explorer-11-ie11/). This makes it seems like GFE laptops are running IE11 in compatibility mode. 

So...can someone with access to a GFE laptop - 

1. Open IE11
1. Navigate to https://www.oprm.va.gov/
1. Open IE settings using the gear icon in the top right corner
1. Select `Compatibility View settings`
1. Uncheck the box `Display intranet sites in Compatibility View`

I used SauceLabs running IE11 and manually added OPRM to the list of sites that should have Compatibility View enabled. Here's what I see -

![image](https://user-images.githubusercontent.com/1915775/76530150-b200ab80-6449-11ea-82bb-c872baed8027.png)

**Comment by ncksllvn** [comment]
In the HTML source of OPRM, I see this line -

![image](https://user-images.githubusercontent.com/1915775/76530709-80d4ab00-644a-11ea-9b95-76ff59107d44.png)

I think that should be changed to -

```
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
```

To align with www.va.gov. I think that is the problem with this site.

**Comment by ncksllvn** [comment]
Okay, reading this [StackOverflow thread](https://stackoverflow.com/questions/6771258/what-does-meta-http-equiv-x-ua-compatible-content-ie-edge-do), I feel confident that meta tag is the issue. It's forcing compatibility view, which is why this issue is unique to this website.

Is there any way we can have the OPRM folks change that meta tag and test it? I don't think we can do it from our inject script, because the JavaScript has already begun execution.

**Comment by KevinMHoffmanUSDS** [comment]
There is! Thank you, @ncksllvn!!! @drorva I'll cc you on a message shortly.

**Comment by ncksllvn** [comment]
I hope that works! That tag and the way it's used here is confusing, but that tag definitely looks suspicious to me and plays into how Compatibility View seems to be rendering things.

**Comment by Cooperjstn** [comment]
Hello everyone, this is Justin from OPRM. I've got it working. I moved the meta tags above the injection and now the site is displaying properly. 

http://vaww.oprm.webdeve.va.gov/

**Comment by Cooperjstn** [comment]
Thank you everyone for your help. We greatly appreciate it.

**Comment by ncksllvn** [comment]
> I moved the meta tags above the injection and now the site is displaying properly.

Hm, so I suppose the meta tag itself was okay; it just needed to be moved higher up on the page before JS begins execution. Amazing, so glad it worked!!

**Comment by ncksllvn** [comment]
Thank you @Cooperjstn!

**Comment by KevinMHoffmanUSDS** [comment]
Thanks for documenting the solution here, @Cooperjstn! Closing this ticket.
